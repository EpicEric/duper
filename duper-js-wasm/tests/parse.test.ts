import { assert, describe, expect, it } from "vitest";
import { parse, DuperValue } from "..";

describe("parse", () => {
  it("parses an empty object", () => {
    const duper = parse("{}");
    assert.instanceOf(duper, DuperValue);
    expect(duper.type).eq("object");
  });

  it("parses into a JSON-stringifiable object", () => {
    const duper = parse(`
      Product({
        product_id: Uuid("1dd7b7aa-515e-405f-85a9-8ac812242609"),
        name: "Wireless Bluetooth Headphones",
        brand: "AudioTech",
        price: Decimal("129.99"),
        dimensions: (18.5, 15.2, 7.8),  // In centimeters
        weight: Kilograms(0.285),
        in_stock: true,
        specifications: {
          battery_life: Duration("30h"),
          noise_cancellation: true,
          connectivity: ["Bluetooth 5.0", "3.5mm Jack"],
        },
        image_thumbnail: Png(b"\\x89PNG\\r\\n\\x1a\\n\\x00\\x00\\x00\\rIHDR\\x00\\x00\\x00\\x64"),
        tags: ["electronics", "audio", "wireless"],
        release_date: Date("2023-11-15"),
        /* Warranty is optional */
        warranty_period: null,
        customer_ratings: {
          latest_review: r#"Absolutely ""astounding""!! 😎"#,
          average: 4.5,
          count: 127,
        },
        created_at: DateTime("2023-11-17T21:50:43+00:00"),
      })
    `);
    assert.instanceOf(duper, DuperValue);
    expect(duper.type).eq("object");
    expect(duper.identifier).eq("Product");
    expect(JSON.stringify(duper)).toMatchSnapshot();
  });

  it("parses into a toString-able array", () => {
    const duper = parse(`
      Foo([b"bar", false, 3.14])
    `);
    assert.instanceOf(duper, DuperValue);
    expect(duper.type).eq("array");
    expect(duper.identifier).eq("Foo");
    expect(duper.toString()).toMatchSnapshot();
  });

  it("doesn't parse invalid Duper values", () => {
    assert.throws(() => parse(``));
    assert.throws(() => parse(`{`));
    assert.throws(() => parse(`]`));
    assert.throws(() => parse(`tru`));
    assert.throws(() => parse(`.618`));
    assert.throws(() => parse(`Something(1, 2)`));
    assert.throws(() => parse(`Iñvalid({})`));
  });
});
