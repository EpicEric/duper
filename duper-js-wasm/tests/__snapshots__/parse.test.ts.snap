// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > parses into a JSON-stringifiable object 1`] = `"{"product_id":"1dd7b7aa-515e-405f-85a9-8ac812242609","name":"Wireless Bluetooth Headphones","brand":"AudioTech","price":"129.99","dimensions":[18.5,15.2,7.8],"weight":0.285,"in_stock":true,"specifications":{"battery_life":"30h","noise_cancellation":true,"connectivity":["Bluetooth 5.0","3.5mm Jack"]},"image_thumbnail":[137,80,78,71,13,10,26,10,0,0,0,13,73,72,68,82,0,0,0,100],"tags":["electronics","audio","wireless"],"release_date":"2023-11-15","warranty_period":null,"customer_ratings":{"latest_review":"Absolutely \\"\\"astounding\\"\\"!! ðŸ˜Ž","average":4.5,"count":127},"created_at":"2023-11-17T21:50:43+00:00"}"`;

exports[`parse > parses into a toString-able array 1`] = `"Foo([b"bar", false, 3.14])"`;
